### PR TITLE
`sort()`, `rsort()` and `usort()` converts an array to list

### PR DESCRIPTION
**Pull request by takaram**
Closes phpstan/phpstan#6467
Closes phpstan/phpstan#10423
Closes phpstan/phpstan#3312

### CODE REVIEW

**Comment by staabm** [comment]
Please have a look at https://github.com/phpstan/phpstan/issues/3312

It seems also fixed by your PR. If so, please add a test

**Comment by takaram** [comment]
Yes, this PR will fix it too. I added a test.
Thank you!

File: src/Analyser/NodeScopeResolver.php [hunk 1]
```diff
@@ -3079,6 +3095,24 @@ static function (?Type $offsetType, Type $valueType, bool $optional) use (&$arra
 		return $arrayType;
 	}
 
+	private function getArraySortFunctionType(Type $type): Type
+	{
+		if (!$type->isArray()->yes()) {
+			return new ErrorType();
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php]: Just return the same `$type` instead

File: src/Analyser/NodeScopeResolver.php [hunk 2]
```diff
@@ -3079,6 +3095,24 @@ static function (?Type $offsetType, Type $valueType, bool $optional) use (&$arra
...
+		}
+
+		$isIterableAtLeastOnce = $type->isIterableAtLeastOnce();
+		if ($isIterableAtLeastOnce->no()) {
+			return new ConstantArrayType([], []);
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php]: Just return the same `$type` instead.

File: src/Analyser/NodeScopeResolver.php [hunk 3]
```diff
@@ -2128,6 +2129,21 @@ static function (): void {
 				);
 			}
 
+			if (
+				$functionReflection !== null
+				&& (
+					(in_array($functionReflection->getName(), ['sort', 'rsort'], true) && count($expr->getArgs()) >= 1)
+					|| ($functionReflection->getName() === 'usort' && count($expr->getArgs()) >= 2)
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php]: We don't need to complicate this by checking that the number of args is `>= 2`. Just check that the first arg is always there and join the condition into a single `in_array`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Otherwise perfect 👍

**Comment by takaram** [comment]
@ondrejmirtes 
Thank you. I fixed the code.

**Comment by takaram** [comment]
https://github.com/phpstan/phpstan-src/actions/runs/7682388532/job/20936712149
This failure looks to be caused by this PR, but I'm not sure why it fails🤔
I need some more investigation

**Comment by takaram** [comment]
> https://github.com/phpstan/phpstan-src/actions/runs/7682388532/job/20936712149

These errors were reported before, but they were in the [baseline](https://github.com/doctrine/persistence/blob/e7cf418cafe83a75f7a7096cc9fa890b201d28f1/phpstan-baseline.neon#L33-L41).
This PR changes the type descriptions, so baseline no longer suppresses the errors.

So I think we can ignore the CI error.

File: src/Analyser/NodeScopeResolver.php (hunk 1 above)
Comment (takaram) [reply src/Analyser/NodeScopeResolver.php]: Fixed

File: src/Analyser/NodeScopeResolver.php (hunk 2 above)
Comment (takaram) [reply src/Analyser/NodeScopeResolver.php]: Fixed

File: src/Analyser/NodeScopeResolver.php (hunk 3 above)
Comment (takaram) [reply src/Analyser/NodeScopeResolver.php]: Fixed

**Comment by ondrejmirtes** [comment]
> So I think we can ignore the CI error.

Yes we can :)

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**